rule bedtools_coverage:
    input:
        windows=re.sub("(\.fasta$|\.fa$)", "100k_windows.bed", config["genome"]),
        genome=re.sub("(\.fasta$|\.fa$)", ".genome", config["genome"]),
        bam="data/merged/{sample}.bam",
        bai="data/merged/{sample}.bam.bai"
        # bam="data/realigned/{sample}.bam",
        # bai="data/realigned/{sample}.bam.bai"
    output:
        "data/readcounts/{sample}.bed"
    log:
        "log/readcounts/{sample}.log"
    shell: """
        samtools view -b -q 20 {input.bam} | \
        bedtools coverage \
            -sorted \
            -nonamecheck \
            -header \
            -g {input.genome} \
            -F 0.5 \
            -a {input.windows} \
            -b stdin \
            > {output} \
            2> {log}
        """